File: src/core/engine.js
import { Grid } from "../structs/grid.js";
import { PARTICLE, CATEGORY } from "../structs/data.js";
import { Renderer } from "../io/renderer.js";
import { InputManager } from "../io/inputManager.js";
import { Debug } from "../io/debug.js";
import { Utility } from "../utils/utility.js";

export class Engine {
  // Dependencies
  #renderer;
  #inputManager;
  #debug;
  // Game states
  #gameWidth;
  #gameHeight;
  #currentGrid;
  #nextGrid;
  #particlesProcessed = new Set();
  // Game loop variables
  #isRunning = false;
  #animationFrameId = null;
  #renderUpdateInterval = 15; // renders ~60 frames per second
  #physicsInterval = 15; // updates physics 40 frames per second
  // Time tracking for FPS
  #lastFrameTime = 0;
  #frameCount = 0;
  #fps = 0;
  // Time tracking for TPS
  #lastPhysicsTime = 0;
  #tickCount = 0;
  #tps = 0;
  // Time tracking for renderer, debug, and UI updates
  #lastRenderUpdateTime = 0;

  constructor(gameWidth, gameHeight, rendererInstance, inputManagerInstance, debugInstance) {
    if (
      !(rendererInstance instanceof Renderer) ||
      !(inputManagerInstance instanceof InputManager) ||
      !(debugInstance instanceof Debug)
    ) {
      throw new Error("Engine constructor requires valid instances!");
    }
    this.#renderer = rendererInstance;
    this.#inputManager = inputManagerInstance;
    this.#debug = debugInstance;
    this.#gameWidth = gameWidth;
    this.#gameHeight = gameHeight;
    this.#currentGrid = new Grid(this.#gameWidth, this.#gameHeight);

    this.#currentGrid.populateGrid(PARTICLE.EMPTY);
    this.#renderer.queueParticles(this.#currentGrid.dirtyParticles, this.#debug.isOverlayEnabled, {
      r: 238,
      g: 148,
      b: 210,
      a: 220,
    });
  }
  start() {
    if (!this.#isRunning) {
      this.#isRunning = true;
      this.#gameLoop();
    }
  }

  stop() {
    this.#isRunning = false;
    cancelAnimationFrame(this.#animationFrameId);
    this.#animationFrameId = null;
  }
  #gameLoop = (timestamp) => {
    if (!this.#isRunning) return;

    // ---! FPS Calculation !---
    this.#frameCount++;
    const delta = timestamp - this.#lastFrameTime;
    if (delta >= 1000) {
      // Update FPS every second; 1000ms = 1s
      this.#fps = (this.#frameCount * 1000) / delta;
      this.#frameCount = 0;
      this.#lastFrameTime = timestamp;
    }

    // ---! Physics Update (TPS) !---
    const physicsDelta = timestamp - this.#lastPhysicsTime;
    if (physicsDelta >= this.#physicsInterval) {
      // Step 1. process physics this frame
      this.#stepPhysics();

      this.#tickCount++;
      this.#lastPhysicsTime = timestamp;
    }

    // ---! Debug UI Update !---
    const renderDelta = timestamp - this.#lastRenderUpdateTime;
    if (renderDelta >= this.#renderUpdateInterval) {
      this.#tps = (this.#tickCount * 1000) / renderDelta;

      // Process input and update UI
      this.#inputManager.processInput(this.#currentGrid, this.#renderer);

      // Update debug display
      this.#debug.updateDisplay(this.#fps, this.#tps);

      // Render this frame
      this.#renderer.renderThisFrame();

      this.#tickCount = 0;
      this.#lastRenderUpdateTime = timestamp;
    }

    this.#animationFrameId = requestAnimationFrame(this.#gameLoop);
  };

  // --------- Handle physics ---------
  #stepPhysics() {
    // Store dirty particles of the previous frame to update this frame
    let particlesToUpdate = Array.from(this.#currentGrid.dirtyParticles);

    // Exit if there's nothing to update
    if (particlesToUpdate.length === 0) return;

    // Clear particles processed array and current grid's dirty particles to initialise for the next part
    this.#particlesProcessed.clear();
    // Step 2. push particles to be rendered for this frame
    this.#renderer.queueParticles(this.#currentGrid.dirtyParticles, this.#debug.isOverlayEnabled, {
      r: 238,
      g: 148,
      b: 210,
      a: 255,
    });
    // Step 3. clear any game data related to this frame
    this.#currentGrid.clearDirty();

    // Shuffle the entire list to randomize the horizontal order
    particlesToUpdate = Utility.shuffleArray(particlesToUpdate);

    // Sort from top to bottom (y-coordinate)
    particlesToUpdate.sort((particleA, particleB) => particleB.position.y - particleA.position.y);

    // Loop through previous dirty particles and update them
    for (const particle of particlesToUpdate) {
      if (!particle) continue; // If particle is null, don't bother

      // If the particle was already processed this frame, don't bother
      if (this.#particlesProcessed.has(particle.index)) return;

      switch (particle.category) {
        case CATEGORY.SOLID:
          // Handle solids
          break;
        case CATEGORY.LIQUID:
          // Handle liquids
          break;
        case CATEGORY.GAS:
          // Handle gases
          break;
        case CATEGORY.SAND:
          // Handle sands
          this.#handleSandPhysics(particle, this.#currentGrid, this.#particlesProcessed);
          break;
        default:
          continue;
      }
    }
  }
  #handleSandPhysics(particle, currentGrid, particlesProcessed) {
    //let directions = JSON.parse(JSON.stringify(particle.reposeDirections));
    let directions = particle.reposeDirections;

    // let directions = [
    //   [{ dx: 0, dy: 1 }],
    //   [
    //     { dx: 1, dy: 4 },
    //     { dx: -1, dy: 4 },
    //   ],
    // ];

    const targetParticle = this.#currentGrid.tryMoveParticle(particle, directions, true, true, true);
    if (targetParticle) {
      this.#particlesProcessed.add(particle);
      this.#particlesProcessed.add(targetParticle);
    }
  }

  // --------- Helper Functions ---------
}

File: src/core/engine.rs
use crate::io::native_renderer::NativeRenderer;
use crate::io::renderer_interface::RendererInterface;
use crate::structs::color::Color;
use crate::structs::input_state::InputState;
use crate::structs::utils::{Offset2, Pixel, Vector2};
use crate::structs::{grid::Grid, particle::Particle};

use rand::seq::SliceRandom;
use rand::thread_rng;

pub struct Engine {
    pub game_width: usize,
    pub game_height: usize,
    pub current_grid: Grid,

    pub brush_radius: i32,
    pub selected_particle_id: u16,
}

impl Engine {
    pub fn new(game_width: usize, game_height: usize, renderer: &mut dyn RendererInterface) -> Engine {
        let mut new_engine = Engine {
            game_width,
            game_height,
            current_grid: Grid::new(game_width as u16, game_height as u16),

            brush_radius: 8,
            selected_particle_id: 300,
        };

        // Populate the grid with empty particles
        const EMPTY_PARTICLE_ID: u16 = 0;
        new_engine.current_grid.populate(EMPTY_PARTICLE_ID);

        // Queue grid to be rendered
        renderer.queue_particles(&new_engine.current_grid.data);

        return new_engine;
    }

    pub fn update(&mut self, renderer: &mut dyn RendererInterface, input_state: &InputState) {
        // Process input
        self.handle_input(input_state);

        let mut particles_to_render: Vec<Particle> = Vec::with_capacity(self.current_grid.dirty_particles.len());

        for &index in &self.current_grid.dirty_particles {
            if let Some(particle) = self.current_grid.data.get(index as usize) {
                particles_to_render.push(particle.clone());
            }
        }

        // Clear dirty particle set for the next frame
        self.current_grid.dirty_particles.clear();

        // Update particle physics
        self.step_physics(&mut particles_to_render);

        // Queue dirty particles
        renderer.queue_particles(&particles_to_render);
    }

    fn handle_input(&mut self, input_state: &InputState) {
        // Paint or erase particles
        if input_state.mouse_left_down || input_state.mouse_right_down {
            let pos: &Vector2<f64> = &input_state.mouse_position;

            let mut particle_id: u16 = self.selected_particle_id;
            if input_state.mouse_right_down {
                particle_id = 0; // We are erasing
            }

            if pos.x >= 0.0 && pos.x < self.game_width as f64 && pos.y >= 0.0 && pos.y < self.game_height as f64 {
                self.current_grid
                    .fill_circle_at(pos.x as i32, pos.y as i32, self.brush_radius, particle_id);
            }
        }
    }

    fn step_physics(&mut self, particles_to_update: &mut Vec<Particle>) {
        // Shuffle to randomize horizontal order
        let mut rng = thread_rng();
        particles_to_update.shuffle(&mut rng);

        // Sort from bottom to top (increasing y)
        particles_to_update.sort_by(|a: &Particle, b: &Particle| a.position.y.cmp(&b.position.y));

        for particle in particles_to_update {
            match particle.category {
                4 => self.handle_sands(particle.index as usize),
                _ => {}
            }
        }
    }

    fn handle_sands(&mut self, particle_index: usize) {
        let direction_groups: Vec<Vec<Offset2<i32>>> = vec![
            vec![Offset2::<i32> { dx: 0, dy: -1 }],
            vec![Offset2::<i32> { dx: 1, dy: -1 }, Offset2::<i32> { dx: -1, dy: -1 }],
        ];

        self.current_grid
            .try_move_particle(particle_index, &direction_groups, true, true, true);
    }
}
